values/dynamic.go

```diff
@@ -0,0 +1,108 @@
+package values
+
+import (
+	"regexp"
+
+	"github.com/influxdata/flux/semantic"
+)
+
+type Dynamic interface {
+	Value
+	Inner() Value
+}
+
+type dynamic struct {
+	inner Value
+}
+
+func NewDynamic(inner Value) Dynamic {
+	return dynamic{inner: inner}
+}
+
+func (d dynamic) Inner() Value {
+	return d.inner
+}
+
+func (d dynamic) Type() semantic.MonoType {
+	return semantic.NewDynamicType()
+}
+
+func (d dynamic) IsNull() bool {
+	return d.inner.IsNull()
+}
+
+func (d dynamic) Str() string {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.String))
+}
+
+func (d dynamic) Bytes() []byte {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Bytes))
+}
+
+func (d dynamic) Int() int64 {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Int))
+}
+
+func (d dynamic) UInt() uint64 {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.UInt))
+}
+
+func (d dynamic) Float() float64 {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Float))
+}
+
+func (d dynamic) Bool() bool {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Bool))
+}
+
+func (d dynamic) Time() Time {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Time))
+}
+
+func (d dynamic) Duration() Duration {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Duration))
+}
+
+func (d dynamic) Regexp() *regexp.Regexp {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Regexp))
+}
+
+func (d dynamic) Array() Array {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Array))
+}
+
+func (d dynamic) Object() Object {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Object))
+}
+
+func (d dynamic) Function() Function {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Function))
+}
+
+func (d dynamic) Dict() Dictionary {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Dictionary))
+}
+
+func (d dynamic) Dynamic() Dynamic {
+	return d
+}
+
+func (d dynamic) Vector() Vector {
+	panic(UnexpectedKind(semantic.Dynamic, semantic.Vector))
+}
+
+func (d dynamic) Equal(v Value) bool {
+	// XXX(sean): Do we need a custom implementation of Equal() for dynamic?
+	// Or are we fine deferring to the inner type?
+	// Seems like a dynamic value should not equal a static value, even if they're
+	// the same under the hood.
```

Comment:
It's an interesting question. I could see this returning `true` if v successfully typechecks as a `Dynamic` and the `d.inner.Equal(v.inner)` is `true`.

I think the thing to consider is when this `Equal` method is used. Do we delegate to this via the `==` binary op? I would expect not, but I don't know.

Reply:
It's probably a good idea to look at some of the other value type impls and what they do for `Equal`. I bet there's a precedent well established by now. I'd look at some scalar types (int, float -- float being the one I'd expect to be more irregular because of precision concerns) as well as array or object, just to see how we compare composite types.

From there, we'd need to make a decision about which model makes more sense: basic or composite.

Reply:
The other impls I saw seem to follow this pattern:

- Check that the types/natures are the same
- If either side of the comparison is null, the values are not equal
- Check that the values are the same

For dynamic, I figured it would be sufficient to just check that both values are dynamic, and defer to the `Equal()` method of the inner type for all of the above checks.

Reply:
Lucky guess on my part then :rofl: 
Sounds great.